src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -519,101 +661,119 @@ void performReplicas(PrimaryOperationRequest por, Tuple<Response, ReplicaRequest
-            }
-            ReplicationState replicationState = new ReplicationState(por, shardIt, primaryResponse.v1(), primaryResponse.v2(), listener, numberOfPendingShardInstances, numberOfUnassignedReplicas);
-            if (numberOfPendingShardInstances == 0) {
-                replicationState.forceFinish();
-                return;
-            }
-            IndexMetaData indexMetaData = observer.observedState().metaData().index(internalRequest.concreteIndex());
-            if (newPrimaryShard != null) {
-                performOnReplica(replicationState, newPrimaryShard, newPrimaryShard.currentNodeId(), indexMetaData);
-            }
+            // one for the primary already done
+            this.totalShards = 1 + numberOfPendingShardInstances + numberOfUnassignedReplicas;
+            this.pending = new AtomicInteger(numberOfPendingShardInstances);
+            this.indexMetaData = observer.observedState().metaData().index(internalRequest.concreteIndex());
+        }
 
-            shardIt.reset(); // reset the iterator
-            while ((shard = shardIt.nextOrNull()) != null) {
-                // if its unassigned, nothing to do here...
-                if (shard.unassigned()) {
-                    continue;
+        /** total shard copies */
+        int totalShards() {
+            return totalShards;
+        }
+
+        /** total successful operations so far */
+        int successful() {
+            return success.get();
+        }
+
+        /** number of pending operations */
+        int pending() {
+            return pending.get();
+        }
+
+        /** start sending current requests to replicas */
+        public void start() {
+            try {
+                if (pending.get() == 0) {
+                    doFinish();
+                    return;
                 }
+                ShardRouting shard;
+                shardIt.reset(); // reset the iterator
+                while ((shard = shardIt.nextOrNull()) != null) {
+                    // if its unassigned, nothing to do here...
+                    if (shard.unassigned()) {
+                        continue;
+                    }
 
-                // if the shard is primary and relocating, add one to the counter since we perform it on the replica as well
-                // (and we already did it on the primary)
-                boolean doOnlyOnRelocating = false;
-                if (shard.primary()) {
-                    if (shard.relocating()) {
-                        doOnlyOnRelocating = true;
+                    // we index on a replica that is initializing as well since we might not have got the event
+                    // yet that it was started. We will get an exception IllegalShardState exception if its not started
+                    // and that's fine, we will ignore it
+                    if (shard.primary()) {
+                        if (originalPrimaryShard.currentNodeId().equals(shard.currentNodeId()) == false) {
```

Comment:
if this return `true` then that would be odd? maybe we should log this and add an assert?